File: app/src/main/java/com/princeparadoxes/watertracker/data/DataModule.java
package com.princeparadoxes.watertracker.data;

import android.app.Application;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.princeparadoxes.watertracker.ApplicationScope;
import com.princeparadoxes.watertracker.data.sp.ProjectPreferenceModule;

import java.io.File;
import java.util.concurrent.TimeUnit;

import dagger.Module;
import dagger.Provides;
import okhttp3.Cache;
import okhttp3.OkHttpClient;

@Module(includes = {ProjectPreferenceModule.class})
public class DataModule {

    static final int DISK_CACHE_SIZE = 50 * 1024 * 1024; // 50MB

    @Provides
    @ApplicationScope
    Gson provideGson() {
        return new GsonBuilder()
//                .registerTypeAdapterFactory(GsonAdapterFactory.create())
//                .serializeNulls()
                .create();
    }

    static OkHttpClient.Builder createOkHttpClient(Application app) {
        // Install an HTTP cache in the application cache directory.
        File cacheDir = new File(app.getCacheDir(), "http");
        Cache cache = new Cache(cacheDir, DISK_CACHE_SIZE);
        return new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .cache(cache);
    }

}

File: app/src/main/java/com/princeparadoxes/watertracker/data/DataModule.kt
package com.princeparadoxes.watertracker.data

import com.princeparadoxes.watertracker.ApplicationScope
import com.princeparadoxes.watertracker.data.repository.DrinkRepository
import com.princeparadoxes.watertracker.data.repository.SettingsRepository
import com.princeparadoxes.watertracker.data.source.db.DatabaseModule
import com.princeparadoxes.watertracker.data.source.db.service.DrinkDatabaseService
import com.princeparadoxes.watertracker.data.source.sp.ProjectPreferenceModule
import com.princeparadoxes.watertracker.data.source.sp.ProjectPreferences
import com.princeparadoxes.watertracker.domain.interactor.DrinkInputPort
import com.princeparadoxes.watertracker.domain.interactor.settings.SettingsInputPort

import dagger.Module
import dagger.Provides

@Module(includes = [(ProjectPreferenceModule::class), (DatabaseModule::class)])
class DataModule {

    @ApplicationScope
    @Provides
    internal fun provideDrinkInputPort(drinkDatabaseService: DrinkDatabaseService,
                                       projectPreferences: ProjectPreferences): DrinkInputPort {
        return DrinkRepository(drinkDatabaseService, projectPreferences)
    }

    @ApplicationScope
    @Provides
    internal fun provideDrinkInputGateway(projectPreferences: ProjectPreferences): SettingsInputPort {
        return SettingsRepository(projectPreferences)
    }

}
